internal/bminventory/inventory.go

```diff
@@ -1500,23 +1500,67 @@ func (b *bareMetalInventory) refreshClusterHosts(ctx context.Context, cluster *c
 		log.WithError(err).Errorf("Failed to set cluster %s majority groups", cluster.ID.String())
 		return common.NewApiError(http.StatusInternalServerError, err)
 	}
+
+	dbCluster, err := common.GetClusterFromDB(tx, *cluster.ID, true)
+	if err != nil {
+		log.WithError(err).Errorf("not refreshing cluster hosts - failed to find cluster %s", *cluster.ID)
+		if errors.Is(err, gorm.ErrRecordNotFound) {
+			return common.NewApiError(http.StatusNotFound, err)
+		}
+		return common.NewApiError(http.StatusInternalServerError, err)
+	}
+
+	dbHosts := make(map[strfmt.UUID]*models.Host)
```

Comment:
Not sure I understand why this map is needed? 
Why not go over the hosts you get from `dbCluster`?

Reply:
That's to keep original code guarantees - in the original code there is a check that protects the data from potential inconsistency between data in the cluster object being updated and data in DB:
https://github.com/openshift/assisted-service/blob/732ff1aabf3b3816f1b01d46393ddc03ae8a128e/internal/bminventory/inventory.go#L1507. 

The equivalent in the new approach is to check existence of host from the updated cluster object in map of hosts coming freshly from DB.
If you think that is not a real problem, I will remove that code.

Reply:
Before returning the cluster to the user there is another call to get a cluster so the user will get whatever the DB has after the commit.
In this case, you don't iterate over whatever user gave you, the code iterate over all the hosts, so unless I miss something this is not really needed. 
There is no comment in the original code explaining why it was done but I think that was related to getting the lasters host.

Reply:
I checked the behavior of the code if `dbCluster.Hosts` is used and it causes issues (`RefreshStatus` returns "no condition passed to run transition RefreshHost from state known" error); object from `cluster.Hosts` has to be used to have expected behavior.

Reply:
it's a bug in the state machine, need to add known to insufficient in https://github.com/openshift/assisted-service/blob/master/internal/host/statemachine.go#L501
HW requirement is changing and the host should not be known anymore.